tests/test_peak_core_and_tile.py

```diff
@@ -0,0 +1,197 @@
+    ], circuit.clk, circuit.reset)
+    tester.reset()
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        for filename in dw_files:
+            shutil.copy(filename, tempdir)
+
+        tester.compile_and_run("verilator",
+                               directory=tempdir,
+                               magma_opts={"coreir_libs": {"float_DW"}},
+                               flags=["-Wno-fatal"])
+
+
+def test_peak_tile_sequence(sequences, dw_files):
+    """
+    Tile level test:
+    * Generates CGRA to extract PE_tile
+    * sets tile_id input based on test app
+    * configures PE_tile based on fixed route (taken from a test app)
+    * similar input driver and output monitor behavior to core test except:
+      * inputs are driven onto the appropriate tile ports based on the fixed
+        route configuration
+      * output is similarly monitored based on the route config
+    """
+    # Use stub CGRA to get PE_tile
+    # TODO: Is there an API to just get a PE_tile?
+    io_sides = IOSide.North | IOSide.East | IOSide.South | IOSide.West
+    interconnect = create_cgra(2, 2, io_sides, num_tracks=3)
+
+    # Use test application to get a configuration
+    # TODO: We should randomize or cover all possible configurations of the
+    # CB/SBs
+    # placement, routing generated from:
+    # netlist = {
+    #     "e0": [("I0", "io2f_16"), ("p0", "data0")],
+    #     "e1": [("I1", "io2f_16"), ("p0", "data1")],
+    #     "e3": [("p0", "alu_res"), ("I2", "f2io_16")],
+    # }
+    # bus = {"e0": 16, "e1": 16, "e3": 16}
+
+    # from archipelago import pnr
+    # placement, routing = pnr(interconnect, (netlist, bus))
+    # route_config = interconnect.get_route_bitstream(routing)
+    # x, y = placement["p0"]
+
+    # placement = {'I0': (1, 0), 'I1': (2, 0), 'I2': (0, 1), 'p0': (1, 1)}
+    # routing = {'e0': [[CB_io2f_16, SB_T0_NORTH_SB_IN_B16, CB_data0]],
+    #            'e1': [[CB_io2f_16, SB_T0_NORTH_SB_IN_B16,
+    #                    SB_T0_WEST_SB_OUT_B16, RMUX_T0_WEST_B16,
+    #                    SB_T0_EAST_SB_IN_B16, CB_data1]],
+    #            'e3': [[CB_alu_res, SB_T0_WEST_SB_OUT_B16, RMUX_T0_WEST_B16, CB_f2io_16]]}
+
+    route_config = [(262401, 0), (459265, 0), (459265, 0), (327937, 2),
+                    (459009, 786432), (459009, 0)]
+    route_config = compress_config_data(route_config)
+    x, y = (1, 1)
+
+    tile_id = x << 8 | y
+    tile = interconnect.tile_circuits[x, y]
+    circuit = tile.circuit()
+
+    def tile_input_driver(tester, value):
+        config_data, a, b, = value
+        for addr, data in config_data:
+            # Convert to tile level address space
+            addr = interconnect.get_config_addr(addr, 0, x, y)
+            tester.configure(addr, data)
+
+        # TODO: We assume these inputs from the test routing app,
+        # this should be done based on the configuration (so we can test
+        # multiple)
+        tester.poke(tester._circuit.SB_T0_NORTH_SB_IN_B16, a)
```

Comment:
For now this is fine. I understand that there is no function to get the actual port name from PnR node yet. I will add it in a future PR.

Reply:
Given the new helper function, you can dynamically figure out which top-level port to poke and expect. This avoids hard-coding ports to poke/expect. You can do something like this. Maybe I'm doing something wrong, but that `circuit.circuit` thing is a little bit verbose to me.

```diff
diff --git a/tests/test_peak_core/test_peak_core_and_tile.py b/tests/test_peak_core/test_peak_core_and_tile.py
index 67764582..15fcdb05 100644
--- a/tests/test_peak_core/test_peak_core_and_tile.py
+++ b/tests/test_peak_core/test_peak_core_and_tile.py
@@ -157,9 +157,11 @@ def test_peak_tile_sequence(sequence, cw_files):
 
     route_config = compress_config_data(route_config)
 
-    tile_id = x << 8 | y
-    tile = interconnect.tile_circuits[x, y]
-    circuit = tile.circuit()
+    circuit = interconnect.circuit()
+
+    input_a = interconnect.get_top_input_port_by_coord(placement["I0"], 16)
+    input_b = interconnect.get_top_input_port_by_coord(placement["I1"], 16)
+    output_port = interconnect.get_top_output_port_by_coord(placement["I2"], 16)
 
     class TileDriver(Driver):
         def lower(self, config_data, a, b, output):
@@ -169,19 +171,18 @@ def test_peak_tile_sequence(sequence, cw_files):
             # TODO: We assume these inputs from the test routing app,
             # this should be done based on the configuration (so we can test
             # multiple)
-            self.tester.circuit.SB_T0_NORTH_SB_IN_B16 = a
-            self.tester.circuit.SB_T2_EAST_SB_IN_B16 = b
+            tester.poke(self.tester.circuit.circuit.interface[input_a], a)
+            tester.poke(self.tester.circuit.circuit.interface[input_b], b)
 
     class TileMonitor(Monitor):
         def observe(self, config_data, a, b, output):
             # TODO: We assume this output from the test routing app,
             # should be based on cofig
-            self.tester.circuit.SB_T0_WEST_SB_OUT_B16.expect(output)
+            tester.expect(self.tester.circuit.circuit.interface[output_port], output)
 
     tester = BasicSequenceTester(circuit, TileDriver(), TileMonitor(),
                                  sequence, circuit.clk, circuit.reset)
     tester.reset()
-    tester.poke(circuit.tile_id, tile_id)
     for addr, data in route_config:
         tester.configure(addr, data)
 

```

Reply:
I think the goal was to lift the test just to the tile level rather than the full interconnect, is possible to add a similar API for the tile level?

Reply:
There is no such functions to create a tile yet:
- Interconnect requires the minimum size to be 2x2
- Tile object can't dump PnR collateral

How about allowing interconnect to be 1x1? In this case although all the ports will be lifted up to the interconnect, it is essentially a tile. And you get the benefit of automatic PnR.

If you think it is a good idea I will work on that now.

Reply:
If 1x1 is easy, then that seems like a decent temporary solution.  I'll talk to other to see it's really necessary to test a tile standalone vs a 1x1.  Is it a one to one mapping from tile to interconnect in the 1x1 case? Or is there any other extra logic that may need to considered as part of the test?

It makes sense that the tile needs PnR information to test it (for understanding the configuration logic), so testing a standalone tile doesn't make sense unless PnR can generate collateral for it (but doing PnR doesn't really make sense for a standalone tile does it?).

Reply:
I guess the interconnect adds the tile_id value, but that seems fine since it's captured in the automatically generated configuration

Reply:
> If 1x1 is easy, then that seems like a decent temporary solution. 
I should be able to get 1x1 working today

> Is it a one to one mapping from tile to interconnect in the 1x1 case?
Yes. All PnR result are offset at (0, 0), so the bitstream applies to both tile level and 1x1 interconnect.

>  (but doing PnR doesn't really make sense for a standalone tile does it?).
Unfortunately there are some routing network inside the tile level, so you have to route to do the test. That being said, if you don't care about routing conflicts (rarely in tile level test), I can write some simple BFS based router that does this.

One major issue is that there is also no clean interface to create a tile. You have to provide switch box etc. If there is a demand to go through that ruote, I can definitely do that.

Please let me know what you think.

Reply:
Is there a different netlist/bus I should use for the 1x1? With the current test (just switched to 1, 1, IOSide.None_ for create_cgra), I get 

```
  >           raise CalledProcessError(retcode, cmd)
  E           subprocess.CalledProcessError: Command '['/Users/lenny/miniconda3/lib/python3.8/site-packages/placer', '/var/folders/hq/hm6_zxqn1jq6xxj_f2fghc7m00
  00gn/T/tmpd98vwz8l/design.layout', '/var/folders/hq/hm6_zxqn1jq6xxj_f2fghc7m0000gn/T/tmpd98vwz8l/design.packed', '/var/folders/hq/hm6_zxqn1jq6xxj_f2fghc7m0000
  gn/T/tmpd98vwz8l/design.place']' died with <Signals.SIGABRT: 6>.

  ../../miniconda3/lib/python3.8/subprocess.py:364: CalledProcessError
  -------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------
  terminate called after throwing an instance of 'std::out_of_range'
    what():  _Map_base::at
```

Reply:
Actually it works if I use `IOSide.North | IOSide.East | IOSide.South | IOSide.West` rather than `IOSide.None_` so it looks like I just need to include the IOs